iree/compiler/Codegen/Interfaces/BufferizationInterfaces.cpp

```diff
@@ -160,6 +162,121 @@ struct DispatchTensorStoreOpInterface
+  // Take a guard before anything else.
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(op);
+
+  // Nothing to do. This op is already bufferized.
+  if (op.hasBufferSemantics()) return success();
+
+  // Ensure op has only tensors. Allow mixed tensor-buffer mode on a per-need
+  // basis.
+  if (!op.hasTensorSemantics())
+    return op->emitError() << "op does not have tensor semantics";
+
+  // New input operands for the cloned op.
+  SmallVector<Value> newInputBuffers;
+  newInputBuffers.reserve(op.getNumInputs());
+  for (OpOperand *opOperand : op.getInputOperands()) {
+    if (op.isScalar(opOperand)) {
+      newInputBuffers.push_back(opOperand->get());
+      continue;
+    }
+    // Input operands are never written to.
+    newInputBuffers.push_back(*state.getBuffer(
+        rewriter, *opOperand,
+        BufferizationState::ForceInPlacability::FORCE_INPLACE));
+  }
+
+  // New output operands for the cloned op.
+  SmallVector<Value> newOutputBuffers;
+  for (OpResult opResult : op->getOpResults()) {
+    SmallVector<OpOperand *> aliasingOpOperands =
+        state.getAnalysisState().getAliasingOpOperand(opResult);
+    assert(aliasingOpOperands.size() == 1 && "expected 1 OpOperand");
+    FailureOr<Value> resultBuffer =
+        state.getBuffer(rewriter, *aliasingOpOperands.front());
+    if (failed(resultBuffer)) return failure();
+    newOutputBuffers.push_back(*resultBuffer);
+  }
+
+  // Merge input/output operands.
+  SmallVector<Value> newOperands = newInputBuffers;
+  newOperands.append(newOutputBuffers.begin(), newOutputBuffers.end());
+
+  // Set insertion point now that potential alloc/dealloc are introduced.
+  rewriter.setInsertionPoint(op);
+  // Clone the op, but use the new operands. Move the existing block into the
+  // new op. Since the new op does not have any tensor results, it does not
+  // return anything.
+  auto newOp = cast<IREE::LinalgExt::LinalgExtOp>(op.cloneWithoutRegions(
+      rewriter, op.getLoc(), /*resultTypes=*/TypeRange{}, newOperands));
+  int64_t numRegions = op->getNumRegions();
+  for (int64_t i = 0; i < numRegions; ++i) {
+    rewriter.inlineRegionBefore(op->getRegion(i), newOp->getRegion(i),
+                                newOp->getRegion(i).begin());
+  }
+
+  // Replace the results of the old op with the new output buffers.
+  bufferization::replaceOpWithBufferizedValues(rewriter, op, newOutputBuffers);
+
+  return success();
+}
+
+/// Bufferization of ops that implement the LinalgExtOp interface. Replace with
+/// a new op that operates entirely on memrefs.
+template <typename OpTy>
+struct LinalgExtOpInterface
+    : public BufferizableOpInterface::ExternalModel<LinalgExtOpInterface<OpTy>,
+                                                    OpTy> {
+  bool bufferizesToMemoryRead(Operation *op, OpOperand &opOperand,
+                              const AnalysisState &state) const {
+    // All operands are read.
+    // TODO: Is this correct?
+    return true;
```

Comment:
I don't know what the exact semantics of `LinalgExtOp` is. Are the output operands read by the operation?

The current implementation may be overly strict. I'm looking for something like `payloadUsesValueFromOperand` but `LinalgExtOp` does not have it.

Reply:
The LinalgExt ops live in their own dispatch (no fusion) at this moment. The output operands may be read by the operation.

If you're looking for `payloadUsesValueFromOperand`, you can assume that all operands are used in the body. I don't know the right term here. The body semantic here means the ops generated in lower_to_loops. E.g.,

https://github.com/google/iree/blob/f3af522bb20c00aa911f5dc1fd671369841f75de/llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/convert_to_loops.mlir#L527-L532

I'm not sure if this helps. This is the logic in IREE old bufferization. It treats LinalgExt ops like Linalg ops, but ignores payloadUsesValueFromOperand

https://github.com/google/iree/blob/f3af522bb20c00aa911f5dc1fd671369841f75de/iree/compiler/Codegen/Common/LinalgBufferizePass.cpp#L673-L730

Reply:
OK that means this implementation is correct.